File: university-study/算法/2022软院码拉松/code/数字消消乐/Main.java
import java.util.ArrayList;
import java.util.Scanner;

public class Main{
    public static void main(String[] args) {
        Scanner sn = new Scanner(System.in);
        int inputValue = sn.nextInt();
        System.out.println(new Main().lastRemaining(inputValue));
    }

    /**
     * @param n 数组长度
     * @param fromLeft ture 从左到右，false从右到左
     * @return
     */
    public int recursion(int n, boolean fromLeft) {
        if (n == 1) return 1;
        // TODO
        ArrayList<Integer> a =new ArrayList<Integer>();
        int number=1;
        for(int num:a){
            num=number;
            number++;
        }
        while(a.size() != 1){
            if(fromLeft){
                for(int i=0;i<a.size();i++){
                    if((i+1)%2==1){
                        a.remove(i);
                        System.out.println(a.size());
                    }
                }
                fromLeft = false;
            }
            else{
                for(int j=a.size()-1;j>=0;j--){
                    if((a.size()-j)%2==1){
                        a.remove(j);
                        System.out.println(a.size());
                    }
                }
                 fromLeft = true;
            }
        }
        return a.get(0);

    }

    public int lastRemaining(int n) {
        return recursion(n, true);
    }
}

File: university-study/算法/2022软院码拉松/code/数字消消乐/main.cpp
#include <iostream>

#include <vector>

using namespace std;

int recursion(int n, bool fromLeft) {
        if (n == 1) return 1;
        // TODO
        vector<int> a;
        int len = n;
        for(int i=1;i<n;i++){
            a[i]=i+1;
        }
        while(len != 1){
            if(fromLeft){
                for(int i=0;i<len;i++){
                    if((i+1)&1){
                        a.erase(a.begin()+i);
                        len--;
                    }
                }
                fromLeft = false;
            }
            else{
                for(int j=len-1;j>=0;j--){
                    if((len-j)&1){
                        a.erase(a.begin()+j);
                        len--;
                    }
                }
                 fromLeft = true;
            }
            for(int i=0;i<len;i++){
                cout<<a[i];
            }
        }
        return a.back();
}

int lastRemaining(int n) {
        return recursion(n,true);
}


int main()
{
int n;
cin>>n;
bool fromLeft = true;
if (n == 1) return 1;
        // TODO
        vector<int> a;
        int len = n;
        for(int i=1;i<n;i++){
            a[i]=i+1;
        }
        while(len != 1){
            if(fromLeft){
                for(int i=0;i<len;i++){
                    if((i+1)&1){
                        a.erase(a.begin()+i);
                        len--;
                    }
                }
                fromLeft = false;
            }
            else{
                for(int j=len-1;j>=0;j--){
                    if((len-j)&1){
                        a.erase(a.begin()+j);
                        len--;
                    }
                }
                 fromLeft = true;
            }
            for(int i=0;i<len;i++){
                cout<<a[i];
            }
        }
//cout<<lastRemaining(n);
return 0;

}
